### ISSUE TITLE
The enforceConst option of the no-magic-numbers rule incorrectly reports a simple for loop

### ISSUE THREAD

**Issue by revolter** [eslint/eslint#10391]
**Tell us about your environment**

* **ESLint Version:** `v4.19.1`
* **Node Version:** `v8.9.4`
* **npm Version:** `6.0.0`

**What parser (default, Babel-ESLint, etc.) are you using?** `default`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    "env": {
        "browser": true,
        "es6": true,
        "greasemonkey": true
    },
    "extends": "eslint:recommended",
    "globals": {
        "min": false
    },
    "rules": {
        "array-callback-return": "error",
        "block-scoped-var": "error",
        "curly": "error",
        "dot-notation": "error",
        "eol-last": [
            "error",
            "always"
        ],
        "eqeqeq": "error",
        "guard-for-in": "error",
        "indent": [
            "error",
            4, {
                "SwitchCase": 1
            }
        ],
        "linebreak-style": [
            "error",
            "unix"
        ],
        "no-case-declarations": "error",
        "no-else-return": "error",
        "no-empty-function": "error",
        "no-eq-null": "error",
        "no-eval": "error",
        "no-extra-bind": "error",
        "no-floating-decimal": "error",
        "no-implicit-coercion": "error",
        "no-implicit-globals": "error",
        "no-implied-eval": "error",
        "no-invalid-this": "error",
        "no-loop-func": "error",
        "no-lone-blocks": "error",
        "no-magic-numbers": [
            "error", {
                "enforceConst": true,
                "ignoreArrayIndexes": true
            }
        ],
        "no-multiple-empty-lines": [
            "error", {
                "max": 1,
                "maxBOF": 0,
                "maxEOF": 1
            }
        ],
        "no-trailing-spaces": "error",
        "quotes": [
            "error",
            "double"
        ],
        "semi": [
            "error",
            "always"
        ],
        "valid-jsdoc": [
            "error", {
                "requireReturn": false
            }
        ]
    }
}
```

</details>
<br>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
(function () {
    var array = [];

    for (let i = 0; i < array.length; i++) {
        array[i] = i;
    }
})();
```

```bash
./node_modules/.bin/eslint index.js
```

**What did you expect to happen?**

Not throw an error.

**What actually happened? Please include the actual, raw output from ESLint.**

```
/Users/revolt/Development/index.js
  4:18  error  Number constants declarations must use 'const'  no-magic-numbers

✖ 1 problem (1 error, 0 warnings)
```

**Comment by platinumazure** [comment]
Hi @revolter, thanks for the issue!

It doesn't seem like we can enforce `const` on for loop declarations, so I can see why it might be good to have ForStatement initializers be exempt from `enforceConst: true`. I'll label this as a bugfix, but am open to suggestions (i.e., if this should be an enhancement, not a bug, we can change the labeling).

Feel free to submit a PR if you like. Thanks!

**Comment by revolter** [comment]
Wow, I didn't expect for it to be a bug, but rather an issue in my configuration. Isn't anybody else using `for` loops and `enforceConst`?!

Regarding your comment, I think it's a bug as I can't see a scenario in which you'd want it to be constant.

It looks like the change should be somewhere around here:

https://github.com/eslint/eslint/blob/465e615188923577e6ca88c559f0c085eb0a2b37/lib/rules/no-magic-numbers.js#L136

right? I might try and fix it myself but I can't guarantee anything since it would be my first contribution to this (awesome) project.

**Comment by g-plane** [comment]
@revolter I can reproduce this problem.

**Comment by not-an-aardvark** [comment]
I think the intent of the rule is that someone would do this:

```js
const LOOP_START_INDEX = 0;
for (let i = LOOP_START_INDEX; i < array.length; i++) {
  // ...
}
```

So I think this is working as designed, although the default configuration enforces code that (I presume) nobody would actually want. I'm not sure the problem is specific to loop variables, so it might not make sense to add an exception there -- instead, maybe it would be better to configure the rule to add `0` to the default `ignore` list.

**Comment by revolter** [comment]
Adding `0` to the `ignore` list causes `let j = 0;` to pass when it shouldn't. So I still think that allowing 0 for a `for` loop's index is the only case that should be ignored for this case.

**Comment by platinumazure** [comment]
I still think that `enforceConst` should not report for ForStatement initializers-- this is literally not possible:

```js
for (const i = 0; i < arr.length; ++i) {
}
```

But I would still expect 0 to be reported as a magic number in this case by default.

If we also want to completely exempt ForStatement initializers from the magic number check itself, I think that should be treated as a separate option.

@not-an-aardvark What do you think? Am I missing something?

**Comment by revolter** [comment]
Can't it check for magic numbers but not for enforce const for `for` loops?

**Comment by not-an-aardvark** [comment]
I think we're straying from the point of the `enforceConst` option. Imagine the code looked like this:

```js
for (let i = 4124; i < arr.length; i++) {
    // ...
}
```

I would argue that the use of `4124` is a code smell, because it's a "magic number" and provides no readable info about what that particular value is for.

The correct fix for the issue would be to do something like this:

```js
const INITIAL_NUM_WIDGETS = 4124; // (e.g.)

for (let i = INITIAL_NUM_WIDGETS; i < arr.length; i++) {
    // ...
}
```

The `enforceConst` option on the `no-magic-numbers` rule controls whether the declaration of `INITIAL_NUM_WIDGETS` must be declared with `const`, or whether `var` is also acceptable. It's not related to the initialization of `i`.

If the purpose of the rule is to give a name to special constants, then I don't think it makes sense to accept a loop iteration variable as the "name" of the constant, because loop iteration variables change their value, so they would not consistently refer to that constant.

For example, the following correction to the original code would not make sense:

```js
// rename `i` to `INITIAL_NUM_WIDGETS`
for (let INITIAL_NUM_WIDGETS = 4124; INITIAL_NUM_WIDGETS < arr.length; INITIAL_NUM_WIDGETS++) {
    // ...
}
```

I would argue that the above code still has the same code smell as the original code, because it's still not clear where the value of `4124` comes from. The name `INITIAL_NUM_WIDGETS` doesn't clarify the situation, because it refers to a value that changes within the loop, rather than always referring to the value `4124`.

**Comment by not-an-aardvark** [comment]
Marking as `evaluating` because discussion is ongoing about whether this is a bug.

**Comment by revolter** [comment]
> I'm not sure the problem is specific to loop variables

No, but loop variables is a very frequent case. I mean, let's be real, everybody writes a for loop starting with `for (var/let i = 0; ...` and no one created or will create a constant for that `0`.

**Comment by revolter** [comment]
I think it should only automatically allow `0` used as the index's initial value in a `for` loop. So it should throw an error for any value different than `0` or, obviously, when not declared in a `for` loop.

**Comment by revolter** [comment]
@not-an-aardvark, Does my last comment make sense?
